### ISSUE TITLE
Watch out for Dataflow classpath limitation

### ISSUE THREAD

**Issue by jean-philippe-martin** [broadinstitute/hellbender#580]
One of our tests (BaseRecalibratorDataflow, on cloud) started failing. It turns out that the culprit is a Dataflow limitation. This is what I got back from the DF team:

*I examined logs of this pipeline on the service and in this case, metadata.items[3] is the pipelineOptions item, whose biggest part is --filesToStage, built from the classpath: it seems you have too many .jar's in classpath, or the jars have too long (absolute) filenames.
It seems that you are using Gradle and all the absolute filenames point deep inside gradle cache directories.
So, as a work-around, you could consider asking Gradle to build a self-contained distribution of your application, put it in a less deep directory, and run that.*

We may run into this problem for other tests as well, so it's good to know about the issue.

**Comment by lbergelson** [comment]
That's a pain... Did they specify how long a path is acceptable?

**Comment by jean-philippe-martin** [comment]
They didn't, but mentioned that setting l4j's debug level to DEBUG will print more information related to what's being sent to the server.

**Comment by pgrosu** [comment]
@jean-philippe-martin Can you show me how I can replicate this error, since I might have a solution to this.

Thanks,
~p

**Comment by pgrosu** [comment]
@jean-philippe-martin It looks like it works fine on the cloud.  I provided the `HELLBENDER_TEST_APIKEY`, and then removed the `cloud` group - from being excluded in `test` - from within `build.gradle` to look like this:
```
test {
    // enable TestNG support (default is JUnit)
    useTestNG{
        excludeGroups 'bucket'
    }
	...
```

And then ran the following integration test, which finished successfully:

```
$ gradle test -Dtest.single=BaseRecalibratorDataflowIntegrationTest --stacktrace
...
    Runtime.totalMemory()=1541406720

BUILD SUCCESSFUL

Total time: 32.277 secs

This build could be faster, please consider using the Gradle Daemon: http://gradle.org/docs/2.4/userguide/gradle_daemon.html
$
```

How are you getting the error to show up?

Thanks,
Paul

**Comment by jean-philippe-martin** [comment]
@pgrosu exactly the way you did, though I was running the test in the IDE. It looks like this depends on the length of the path, so perhaps the difference is the folder names we're using?

**Comment by pgrosu** [comment]
@jean-philippe-martin Below is my exact folder path I ran it from:

```
$ pwd
/home/pgrosu/me/gg_hellbender_broad_institute/hellbender
$ pwd | wc -m
57
$

```
Give it a try it from the cli - and maybe higher up in the path - and see if that makes a differentce - mine is 57 characters into the path.  How many characters is your path that you run it from?  I can try the same thing and see what happens, and then adjust the `build.gradle` file accordingly.

Thanks,
Paul

**Comment by jean-philippe-martin** [comment]
PWD has nothing to do with the length of the classpath, it uses absolute paths.

**Comment by jean-philippe-martin** [comment]
One way to mitigate this problem is to set the GRADLE_USER_HOME environment variable to move the gradle cache onto a shorter path. This solved the problem for me for BaseRecalibratorDataflow. 
In my case I saw about over 100 jars listed.

**Comment by lbergelson** [comment]
Should we bring this up with the dataflow team.  This seems like a major problem if you have a lot of dependencies like we do now.

**Comment by jean-philippe-martin** [comment]
@lbergelson done.

**Comment by jean-philippe-martin** [comment]
Do we know if it's possible to set up IntelliJ to build a far jar and run off of that, instead of many jars? That would slow us down but at least go around this problem.

**Comment by pgrosu** [comment]
@jean-philippe-martin When I build a fatJar manually, and then run just using that, it seem to throw an exception :(

```
$ gradle fatJar
$ java -jar build/libs/hellbender-all-GATK.4.alpha-380-g9f167ee-SNAPSHOT.jar BaseRecalibratorDataflow  -R src/test/resources/org/broadinstitute/hellbender/tools/BQSR/human_b36_both.chr1_1k.fasta -I src/test/resources/org/broadinstitute/hellbender/tools/BQSR/originalQuals.1kg.chr1.1-1K.1RG.dictFix.bam --secret client-secrets.json -OQ -knownSites src/test/resources/org/broadinstitute/hellbender/tools/BQSR/dbsnp_132.b36.excluding_sites_after_129.chr1_1k.vcf --RECAL_TABLE_FILE walktest.tmp -sortAllCols
...
java.lang.RuntimeException: java.lang.NoSuchMethodError: com.google.common.base.Stopwatch.createStarted()Lcom/google/common/base/Stopwatch;
        at com.google.cloud.dataflow.sdk.Pipeline.run(Pipeline.java:155)
        at org.broadinstitute.hellbender.engine.dataflow.DataflowCommandLineProgram.runPipeline(DataflowCommandLineProgram.java:135)
        at org.broadinstitute.hellbender.engine.dataflow.DataflowCommandLineProgram.doWork(DataflowCommandLineProgram.java:101)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.runTool(CommandLineProgram.java:97)
        at org.broadinstitute.hellbender.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:150)
        at org.broadinstitute.hellbender.Main.instanceMain(Main.java:71)
        at org.broadinstitute.hellbender.Main.main(Main.java:86)
Caused by: java.lang.NoSuchMethodError: com.google.common.base.Stopwatch.createStarted()Lcom/google/common/base/Stopwatch;
        at org.broadinstitute.hellbender.dev.pipelines.bqsr.BaseRecalibratorDataflowUtils$4.startBundle(BaseRecalibratorDataflowUtils.java:223)
$
```
~p
```

**Comment by lbergelson** [comment]
That's not good. We'll have to fix that.

**Comment by jean-philippe-martin** [comment]
@lbergelson I fixed pgrosu's problem, that was #637.

**Comment by jean-philippe-martin** [comment]
Good news! The Dataflow team report that they have fixed this issue by raising the limit 20 fold.
